### PR TITLE
feat: add C implementation for `math/base/special/atan2`

### LINKED ISSUE #1923: [RFC]: Add C implementation for `@stdlib/math/base/special/atan2`
### Description

This RFC proposes adding C implementation for [@stdlib/math/base/special/atan2](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/atan2).

### Related Issues

Related issues #649 

### Questions

No.

### Other

The `atan2` package is dependent on :-
`isinfinite` , `copysign`, `signbit`, `isnan`, `atan`, `pinf` and `pi`, and all of these packages are already implemented in C.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by Pushpendra766**
Resolves # .

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds C implementation for [math/base/special/atan2](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/atan2)
     `double out = stdlib_base_atan2( 2.0, 2.0 ); `

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #1923 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/special/atan2/benchmark/benchmark.native.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	var x;
+	var y;
+	var z;
+	var i;
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		x = ( randu()*100.0 ) - 0.0;
+		y = ( randu()*100.0 ) - 0.0;
+		z = Math.atan2( x, y ); // eslint-disable-line stdlib/no-builtin-math
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/atan2/benchmark/benchmark.native.js]:
```suggestion
		z = atan2( y, x );
```

File: lib/node_modules/@stdlib/math/base/special/atan2/README.md [hunk 2]
```diff
@@ -86,6 +86,98 @@ for ( i = 0; i < 100; i++ ) {
...
+
+<section class="usage">
+
+### Usage
+
+```c
+#include "stdlib/math/base/special/atan2.h"
+```
+
+#### stdlib_base_atan2( x )
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/atan2/README.md]:
```suggestion
#### stdlib_base_atan2( y, x )
```

File: lib/node_modules/@stdlib/math/base/special/atan2/README.md [hunk 3]
```diff
@@ -86,6 +86,98 @@ for ( i = 0; i < 100; i++ ) {
...
+double out = stdlib_base_atan2( 2.0, 2.0 ); 
+// returns ~0.785
+
+out = stdlib_base_atan2( 6.0, 2.0 );
+// returns ~1.249
+```
+
+The function accepts the following arguments:
+
+-   **x**: `[in] double` input value.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/atan2/README.md]: Need to document both parameters here.

File: lib/node_modules/@stdlib/math/base/special/atan2/README.md [hunk 4]
```diff
@@ -86,6 +86,98 @@ for ( i = 0; i < 100; i++ ) {
...
+
+```c
+double stdlib_base_atan2( const double x );
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/atan2/README.md]:
```suggestion
double stdlib_base_atan2( const double y, const double x );
```

File: lib/node_modules/@stdlib/math/base/special/atan2/lib/native.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+*/
+
+'use strict';
+
+// MODULES //
+
+var addon = require( './../src/addon.node' );
+
+/**
+* Computes the arctangent of a double-precision floating-point number.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/atan2/lib/native.js]:
```suggestion
* Computes the angle in the plane (in radians) between the positive x-axis and the ray from `(0,0)` to the point `(x,y)`.
```

File: lib/node_modules/@stdlib/math/base/special/atan2/lib/native.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+*
+* @private
+* @param {number} y - input value
+* @param {number} x - input value
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/atan2/lib/native.js]:
```suggestion
* @param {number} y - `y` coordinate
* @param {number} x - `x` coordinate
```

File: lib/node_modules/@stdlib/math/base/special/atan2/lib/native.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+* @returns {number} arctangent (in radians)
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/atan2/lib/native.js]:
```suggestion
* @returns {number} angle (in radians)
```

File: lib/node_modules/@stdlib/math/base/special/atan2/src/main.c [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+* // returns NaN
+*
+* @example
+* double v = stdlib_base_atan2( NaN, 2.0 );
+* // returns NaN
+*/
+
+double stdlib_base_atan2( const double y, const double x ) {
+    double q;
+    if(stdlib_base_is_nan(x) || stdlib_base_is_nan(y)){
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/atan2/src/main.c]: Please add spaces around parentheses here and below to match our code style conventions.

**Review by Planeshifter — CHANGES_REQUESTED** [review]
Thanks for working on this! I left an initial set of comments and suggestions that need to be addressed.

**Comment by Pushpendra766** [comment]
> /stdlib update-copyright-years

should I change it to 2024 ?

**Comment by Planeshifter** [comment]
@Pushpendra766 The `/stdlib update-copyright-years` command automatically has our stdlib-bot update the copyright years for added files, see: https://github.com/stdlib-js/stdlib/pull/1981/commits/5c3356e8164935be474d510421a4b3a56e49bb2b. Just pull down the changes, no reason for you to manually change them.

**Comment by Pushpendra766** [comment]
> @Pushpendra766 The `/stdlib update-copyright-years` command automatically has our stdlib-bot update the copyright years for added files, see: [5c3356e](https://github.com/stdlib-js/stdlib/commit/5c3356e8164935be474d510421a4b3a56e49bb2b). Just pull down the changes, no reason for you to manually change them.

Okay, thanks.
I have done the suggested changes. I hope it will fix the formatting and documentation issues.

File: lib/node_modules/@stdlib/math/base/special/atan2/src/main.c [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+*
+* @example
+* double v = stdlib_base_atan2( 3.0, NaN );
+* // returns NaN
+*
+* @example
+* double v = stdlib_base_atan2( NaN, 2.0 );
+* // returns NaN
+*/
+
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/atan2/src/main.c]:
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/atan2/src/main.c [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+#include <stdint.h>
+
+// MAIN //
+
+/**
+* Computes the angle in the plane (in radians) between the positive x-axis and the ray from `(0,0)` to the point `(x,y)`.
+*
+* @param {number} y - `y` coordinate
+* @param {number} x - `x` coordinate
+* @returns {number} angle (in radians)
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/atan2/src/main.c]:
```suggestion
* @param y    `y` coordinate
* @param x    `x` coordinate
* @return       angle (in radians)
```

File: lib/node_modules/@stdlib/math/base/special/atan2/src/main.c [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    }
+    q = stdlib_base_atan( y / x );
+    if( x < 0.0 ){
+        if( q <= 0.0 ){
+            return q + STDLIB_CONSTANT_FLOAT64_PI;
+        }
+        return q - STDLIB_CONSTANT_FLOAT64_PI;
+    }
+    return q;
+}
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/atan2/src/main.c]:
```suggestion
    if ( stdlib_base_is_nan( x ) || stdlib_base_is_nan( y ) ) {
        return 0.0 / 0.0;
    }
    if ( stdlib_base_is_infinite( x ) ) {
        if ( x == STDLIB_CONSTANT_FLOAT64_PI ) {
            if ( stdlib_base_is_infinite( y ) ) {
                return stdlib_base_copysign( STDLIB_CONSTANT_FLOAT64_PI / 4.0, y );
            }
            return stdlib_base_copysign( 0.0, y );
        }
        // Case: x is -Infinity
        if ( stdlib_base_is_infinite( y ) ) {
            return stdlib_base_copysign( 3.0 * STDLIB_CONSTANT_FLOAT64_PI / 4.0, y );
        }
        return stdlib_base_copysign( STDLIB_CONSTANT_FLOAT64_PI, y );
    }
    if ( stdlib_base_is_infinite( y ) ) {
        return stdlib_base_copysign( STDLIB_CONSTANT_FLOAT64_PI / 2.0, y );
    }
    if ( y == 0.0 ) {
        if ( x >= 0.0 && !stdlib_base_float64_signbit( x ) ) {
            return stdlib_base_copysign( 0.0, y );
        }
        return stdlib_base_copysign( STDLIB_CONSTANT_FLOAT64_PI, y );
    }
    if ( x == 0.0 ) {
        return stdlib_base_copysign( STDLIB_CONSTANT_FLOAT64_PI / 2.0, y );
    }
    q = stdlib_base_atan( y / x );
    if ( x < 0.0 ) {
        if ( q <= 0.0 ) {
            return q + STDLIB_CONSTANT_FLOAT64_PI;
        }
        return q - STDLIB_CONSTANT_FLOAT64_PI;
    }
    return q;
}
```

**Review by Planeshifter — CHANGES_REQUESTED** [review]
CI is currently failing, please take a look. I think the `manifest.json` file is missing used packages in the `examples` and `benchmark` configurations, namely `random/base/randu`, but looks like we also have a problem with the implementation since the tests fail, although at first glance, it looks fine to me and matching the JS implementation.

**Comment by Pushpendra766** [comment]
> CI is currently failing, please take a look. I think the `manifest.json` file is missing used packages in the `examples` and `benchmark` configurations, namely `random/base/randu`, but looks like we also have a problem with the implementation since the tests fail, although at first glance, it looks fine to me and matching the JS implementation.

`random/base/randu` is not a dependency for atan2, as far as I know.
Also, I checked the implementation thoroughly. It matches the JavaScript implementation of the function.

File: lib/node_modules/@stdlib/math/base/special/atan2/README.md [hunk 12]
```diff
@@ -86,6 +86,99 @@ for ( i = 0; i < 100; i++ ) {
...
+
+### Usage
+
+```c
+#include "stdlib/math/base/special/atan2.h"
+```
+
+#### stdlib_base_atan2( y, x )
+
+Compute the angle in the plane (in radians) between the positive x-axis and the ray from `(0,0)` to the point `(x,y)`.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/README.md]:
```suggestion
Computes the angle in the plane (in radians) between the positive x-axis and the ray from `(0,0)` to the point `(x,y)`.
```

Comment (Pranavchiku) [reply lib/node_modules/@stdlib/math/base/special/atan2/README.md]: Please do not commit it, I'll apply every change in batch.

File: lib/node_modules/@stdlib/math/base/special/atan2/README.md [hunk 13]
```diff
@@ -86,6 +86,99 @@ for ( i = 0; i < 100; i++ ) {
...
+
+<!-- C API usage examples. -->
+
+<section class="examples">
+
+### Examples
+
+```c
+#include "stdlib/math/base/special/atan2.h"
+#include "stdlib/random/base/randu"
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/README.md:157]:
```suggestion
#include "stdlib/random/base/randu.h"
```

File: lib/node_modules/@stdlib/math/base/special/atan2/README.md [hunk 14]
```diff
@@ -86,6 +86,99 @@ for ( i = 0; i < 100; i++ ) {
...
+    double x;
+    int i;
+
+    for ( i = 0; i < 100; i++ ) {
+        y = stdlib_base_random_randu(100);
+        x = stdlib_base_random_randu(100);
+        printf("y: %.4f, \t x: %.4f, \t atan2(y,x): %.4f\n", y, x, stdlib_base_atan2(y, x));
+    }
+    return 0;
+}
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/README.md]:
```suggestion
#include "stdlib/math/base/special/atan2.h"
#include <stdlib.h>
#include <stdio.h>

int main( void ) {
    double y;
    double x;
    double v;
    int i;

    for ( i = 0; i < 100; i++ ) {
        y = ( ( (double)rand() / (double)RAND_MAX ) * 100.0 );
        x = ( ( (double)rand() / (double)RAND_MAX ) * 100.0 );
        v = stdlib_base_atan2( y, x );
        printf( "atan2(%lf, %lf) = %lf\n", y, x, v );
    }
    return 0;
}
```

File: lib/node_modules/@stdlib/math/base/special/atan2/benchmark/c/native/benchmark.c [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+/**
+* Benchmark `atan2`.
+*/
+#include <stdlib.h>
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/benchmark/c/native/benchmark.c:23]:
```suggestion
#include "stdlib/math/base/special/atan2.h"
#include <stdlib.h>
```

File: lib/node_modules/@stdlib/math/base/special/atan2/benchmark/c/native/benchmark.c [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+	double y;
+	double z;
+	double t;
+	int i;
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		x = ( 100.0*rand_double() ) - 0.0;
+		y = ( 100.0*rand_double() ) - 0.0;
+		z = atan2( x, y );
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/benchmark/c/native/benchmark.c]:
```suggestion
		z = stdlib_base_atan2( y, x );
```

File: lib/node_modules/@stdlib/math/base/special/atan2/benchmark/c/native/benchmark.c [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+int main( void ) {
+	double elapsed;
+	int i;
+
+	// Use the current time to seed the random number generator:
+	srand( time( NULL ) );
+
+	print_version();
+	for ( i = 0; i < REPEATS; i++ ) {
+		printf( "# c::%s\n", NAME );
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/benchmark/c/native/benchmark.c]:
```suggestion
		printf( "# c::native::%s\n", NAME );
```

File: lib/node_modules/@stdlib/math/base/special/atan2/examples/c/example.c [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+    double x;
+    int i;
+
+    for ( i = 0; i < 100; i++ ) {
+        y = stdlib_base_random_randu(100);
+        x = stdlib_base_random_randu(100);
+        printf("y: %.4f, \t x: %.4f, \t atan2(y,x): %.4f\n", y, x, stdlib_base_atan2(y, x));
+    }
+    return 0;
+}
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/examples/c/example.c]:
```suggestion
#include "stdlib/math/base/special/atan2.h"
#include <stdlib.h>
#include <stdio.h>

int main( void ) {
	double y;
	double x;
	double v;
	int i;
	for ( i = 0; i < 100; i++ ) {
		y = ( ( (double)rand() / (double)RAND_MAX ) * 100.0 );
		x = ( ( (double)rand() / (double)RAND_MAX ) * 100.0 );
		v = stdlib_base_atan2( y, x );
		printf( "atan2(%lf, %lf) = %lf\n", y, x, v );
	}
	return 0;
}
```

File: lib/node_modules/@stdlib/math/base/special/atan2/include/stdlib/math/base/special/atan2.h [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+
+/*
+* If C++, prevent name mangling so that the compiler emits a binary file having undecorated names, thus mirroring the behavior of a C compiler.
+*/
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+* Computes the arctangent of a double-precision floating-point number.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/include/stdlib/math/base/special/atan2.h]:
```suggestion
* Computes the angle in the plane (in radians) between the positive x-axis and the ray from `(0,0)` to the point `(x,y)`.
```

File: lib/node_modules/@stdlib/math/base/special/atan2/manifest.json [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+      "src": [
+        "./src/main.c"
+      ],
+      "include": [
+        "./include"
+      ],
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/napi/unary",
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/manifest.json]:
```suggestion
        "@stdlib/math/base/napi/binary",
```

File: lib/node_modules/@stdlib/math/base/special/atan2/src/addon.c [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "stdlib/math/base/special/atan2.h"
+#include "stdlib/math/base/napi/unary.h"
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/src/addon.c]:
```suggestion
#include "stdlib/math/base/napi/binary.h"
```

File: lib/node_modules/@stdlib/math/base/special/atan2/src/addon.c [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+
+// cppcheck-suppress shadowFunction
+STDLIB_MATH_BASE_NAPI_MODULE_D_D( stdlib_base_atan2 )
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/src/addon.c]:
```suggestion
STDLIB_MATH_BASE_NAPI_MODULE_DD_D( stdlib_base_atan2 )
```

File: lib/node_modules/@stdlib/math/base/special/atan2/src/main.c [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+#include <stdint.h>
+
+// MAIN //
+
+/**
+* Computes the angle in the plane (in radians) between the positive x-axis and the ray from `(0,0)` to the point `(x,y)`.
+*
+* @param y    `y` coordinate
+* @param x    `x` coordinate
+* @return       angle (in radians)
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/src/main.c]:
```suggestion
* @param x    `x` coordinate
* @return      angle (in radians)
```

File: lib/node_modules/@stdlib/math/base/special/atan2/src/main.c [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+        return stdlib_base_copysign( STDLIB_CONSTANT_FLOAT64_PI / 2.0, y );
+    }
+    q = stdlib_base_atan( y / x );
+    if ( x < 0.0 ) {
+        if ( q <= 0.0 ) {
+            return q + STDLIB_CONSTANT_FLOAT64_PI;
+        }
+        return q - STDLIB_CONSTANT_FLOAT64_PI;
+    }
+    return q;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/src/main.c]:
```suggestion
	double q;
	if ( stdlib_base_is_nan( x ) |	dlib_base_is_nan( y ) ) {
		return 0.0 / 0.0; // NaN
	}
	if ( stdlib_base_is_infinite( x ) ) {
		if ( x == STDLIB_CONSTANT_FLOAT64_PINF ) {
			if ( stdlib_base_is_infinite( y ) ) {
				return stdlib_base_copysign( STDLIB_CONSTANT_FLOAT64_PI / 4.0, y );
			}
			return stdlib_base_copysign( 0.0, y );
		}
		// Case: x is -Infinity
		if ( stdlib_base_is_infinite( y ) ) {
			return stdlib_base_copysign( 3.0 * STDLIB_CONSTANT_FLOAT64_PI / 4.0, y );
		}
		return stdlib_base_copysign( STDLIB_CONSTANT_FLOAT64_PI, y );
	}
	if ( stdlib_base_is_infinite( y ) ) {
		return stdlib_base_copysign( STDLIB_CONSTANT_FLOAT64_PI / 2.0, y );
	}
	if ( y == 0.0 ) {
		if ( x >= 0.0 && !stdlib_base_float64_signbit( x ) ) {
			return stdlib_base_copysign( 0.0, y );
		}
		return stdlib_base_copysign( STDLIB_CONSTANT_FLOAT64_PI, y );
	}
	if ( x == 0.0 ) {
		return stdlib_base_copysign( STDLIB_CONSTANT_FLOAT64_PI / 2.0, y );
	}
	q = stdlib_base_atan( y / x );
	if ( x < 0.0 ) {
		if ( q <= 0.0 ) {
			return q + STDLIB_CONSTANT_FLOAT64_PI;
		}
		return q - STDLIB_CONSTANT_FLOAT64_PI;
	}
	return q;
```

Comment (Pranavchiku) [reply lib/node_modules/@stdlib/math/base/special/atan2/src/main.c]: TAB indentation is required

File: lib/node_modules/@stdlib/math/base/special/atan2/src/main.c [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+* double v = stdlib_base_atan2( 3.0, NaN );
+* // returns NaN
+*
+* @example
+* double v = stdlib_base_atan2( NaN, 2.0 );
+* // returns NaN
+*/
+double stdlib_base_atan2( const double y, const double x ) {
+	double q;
+	if ( stdlib_base_is_nan( x ) |	dlib_base_is_nan( y ) ) {
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/src/main.c]:
```suggestion
	if ( stdlib_base_is_nan( x ) |	stdlib_base_is_nan( y ) ) {
```

File: lib/node_modules/@stdlib/math/base/special/atan2/src/main.c [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+* double v = stdlib_base_atan2( 3.0, NaN );
+* // returns NaN
+*
+* @example
+* double v = stdlib_base_atan2( NaN, 2.0 );
+* // returns NaN
+*/
+double stdlib_base_atan2( const double y, const double x ) {
+	double q;
+	if ( stdlib_base_is_nan( x ) |	stdlib_base_is_nan( y ) ) {
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/atan2/src/main.c]:
```suggestion
	if ( stdlib_base_is_nan( x ) ||	stdlib_base_is_nan( y ) ) {
```

**Review by Pranavchiku — APPROVED** [review]
Tested locally, everything works well. Thanks @Pushpendra766 !

**Review by Planeshifter — APPROVED** [review]

**Pull request merged**